tests/test_sqlalchemy_bulk.py

```diff
@@ -0,0 +1,185 @@
+        |San Francisco|4      |5      |0.25|1994-11-27|
+        |San Francisco|43     |57     |0   |1994-11-29|
+        |Hayward      |37     |54     |None|1994-11-29|
+        +-------------+-------+-------+----+----------+
+        """
+    )
+
+
+def test_render_fail(engine: Engine) -> None:
+    sample_lo_temp = 4
+    with Session(engine) as session, session.begin():
+        session.add_all((
+            Weather(city='San Francisco', temp_lo=4, temp_hi=5, prcp=0.25, date=date(1994, 11, 27)),
+            Weather(city='San Francisco', temp_lo=-1, temp_hi=3, prcp=0.2, date=date(1994, 11, 20)),
+            Weather(city='Hayward', temp_lo=37, temp_hi=54, prcp=None, date=date(1994, 11, 29)),
+        ))
+    with ShouldAssert(
+        dedent("""\
+            sequence not as expected:
+            
+            same:
+            [{'city': 'San Francisco',
+              'date': datetime.date(1994, 11, 27),
+              'prcp': 0.25,
+              'temp_hi': 5,
+              'temp_lo': 4}]
+            
+            expected:
+            [{'city': 'San Francisco',
+              'date': datetime.date(1994, 11, 29),
+              'prcp': 0,
+              'temp_hi': 57,
+              'temp_lo': 43},
+             {'city': 'Hayward',
+              'date': datetime.date(1994, 11, 29),
+              'prcp': None,
+              'temp_hi': 54,
+              'temp_lo': 37}]
+            
+            actual:
+            [{'city': 'San Francisco',
+              'date': datetime.date(1994, 11, 20),
+              'prcp': 0.2,
+              'temp_hi': 3,
+              'temp_lo': -1},
+             {'city': 'Hayward',
+              'date': datetime.date(1994, 11, 29),
+              'prcp': None,
+              'temp_hi': 54,
+              'temp_lo': 37}]
+            
+            While comparing [1]: dict not as expected:
+            
+            same:
+            ['city']
+            
+            values differ:
+            'date': datetime.date(1994, 11, 29) (expected) != datetime.date(1994, 11, 20) (actual)
+            'prcp': 0 (expected) != 0.2 (actual)
+            'temp_hi': 57 (expected) != 3 (actual)
+            'temp_lo': 43 (expected) != -1 (actual)
+            
+            While comparing [1]['date']: datetime.date(1994, 11, 29) (expected) != datetime.date(1994, 11, 20) (actual)
+            
+            While comparing [1]['prcp']: 0 (expected) != 0.2 (actual)
+            
+            Should be:
+            +-------------+----------------+-------+----+----------+
+            |city         |temp_lo         |temp_hi|prcp|date      |
+            +-------------+----------------+-------+----+----------+
+            |San Francisco|4               |5      |0.25|1994-11-27|
+            |San Francisco|-1              |3      |0.2 |1994-11-20|
+            |Hayward      |37              |54     |None|1994-11-29|
+            +-------------+----------------+-------+----+----------+
```

Comment:
It would be nice to also see a unified diff of this table view, to make it easier to see which row(s) are mismatching.

Would require the same column widths on both sides -- I guess the way to do it would be to render both expected and actual sides of the comparison with max column width of either side?

Reply:
Yep, not the most efficient, but the code ended up being more graceful than expected with no changes to the stuff provided  by `chide` 🎉